knowledge-base/adr/0031-resilience-options.md

```diff
@@ -0,0 +1,257 @@
+- The timeout in the `HttpRequestConfigBase` is for all http calls to the target system and passed to axios.
+
+The settings are passed in the following way:
+
+```js
+myApi.getAll().timeout(20).execute({ enableCircuitBreaker: true, timeout: 10 });
+executeHttpRequest(
+  { enableCircuitBreaker: true, timeout: 10 },
+  { timeout: 20 }
+);
+```
+
+So we have something for circuit breaker and timeout.
+
+|     Option      | On target | On BTP | Default target | Default BTP |
+| :-------------: | :-------: | :----: | :------------: | :---------: |
+| circuit breaker |    ❌     |   ✅   |      n.a.      |  enabled   |
+|     timeout     |    ✅     |   ✅   |    enabled     |   enabled   |
+|      retry      |    ❌     |   ❌   |      n.a.      |    n.a.     |
+|   rate limit    |    ❌     |   ❌   |      n.a.      |    n.a.     |
+|   bulk limit    |    ❌     |   ❌   |      n.a.      |    n.a.     |
+
+## Decision
+
+### SDK vs. User
+
+There are two standpoints:
+- User needs to implement it
+- SDK implements it
+
+Arguments in the discussion
+- flexible middle ware approach implemented by the user more flexible
+- You could make mistakes: multi-tenant circuit breaker or retry if breaker is open
+- Configuration on a per request wanted
+
+## Decision
+
+- Rate limit and bulk limit we will not do since -> no user request.
+- Step 1: Circuit breaker is added and made tenant aware 
+- Step 2: Retry is added and excluded if circuit breaker is open.
+- Step 3 (Optional): Make resilience globally configurable for all requests.
+Do this on demand or after customer feedback.
+- Default for circuit breaker is on.
+- Default for retry is off since this seems a bigger behavior change
+
+|     Option      | On target | On BTP | Default target | Default BTP | Remarks |
+| :-------------: | :-------: | :----: | :------------: | :---------: | --- |
+| circuit breaker |    ✅     |   ✅   |    enabled     |   enabled   | tenant aware |
+|     timeout     |    ✅     |   ✅   |    enabled     |   enabled   |  |
+|      retry      |    ✅     |   ✅   |    disabled    |   disabled  | no retry: circuit breaker open and 401,403 |
+|   rate limit    |    ❌     |   ❌   |      n.a.      |    n.a.     | |
+|   bulk limit    |    ❌     |   ❌   |      n.a.      |    n.a.     | |
+
+
+
+### Variant A - Opinionated
+
+- We pick an implementation and only provide options
+- For retry, we would use [async retry](https://www.npmjs.com/package/async-retry)
+- For circuit breaker we would use [opossum](https://www.npmjs.com/package/opossum)
+
+This determines the options.
+The API would look like:
+
+```ts
+myApi
+  .getAll()   
+  .timeout(20) //deprecate 
+  .resiliencec({
+      timeout: 10,
+      circuitBreaker: true,   // CircuitBreakerOptions | undefined | true
+      retry: true             // RetryOptions | undefined | true
+  })
+  .execute({ 
+      enableCircuitBreaker: true, timeout: 10, //deprecate
+      destinationName: 'my-dest'
+  });
+executeHttpRequest({
```

Comment:
```suggestion

executeHttpRequest({
```

Reply:
There is an empty line above the function call 😉